### PR TITLE
scheduler: performance improvement on PodAffinity

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind design
/sig scheduling
/assign @bsalamat

**What this PR does / why we need it**:

This PR tries to eliminate unnecessary Lock/Unlock in the logic of InterPodAffinity priorities. By replacing them with atomic AddInt64 can significantly improve the performance of:

- **_Hard_** PodAffinity (2.2X performance improvement, see below)

    - Before
    
        ```
        BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8               1000         284,892,826 ns/op
        BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8               1000         285,320,757 ns/op
        BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8               1000         287,498,053 ns/op
        ```
    
    - After (with this PR)
    
        ```
        BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8               1000         127,001,991 ns/op
        BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8               1000         129,263,078 ns/op
        BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8               1000         125,813,803 ns/op
        ```

- **_Soft_** PodAffinity/PodAntiAffinity (Can be inferred from the code and benchmark result of Hard PodAffinity. We can add the benchmark tests if necessary)

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

Above test result is run at a Baremetal machine with 8 core cpus and 32GB memory.

**Does this PR introduce a user-facing change?**:

```release-note
2X performance improvement on both required and preferred PodAffinity.
```

### CODE REVIEW

**Review by wgliang — APPROVED** [review]
Cool, LGTM.
Will leave /lgtm to @bsalamat.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go
Comment (ravisantoshgudimetla) [inline pkg/scheduler/algorithm/priorities/interpod_affinity.go]: I guess, we might have lost the float scores by now. I believe, we tried something similar in the past and it affected correctness. The one thing, I am curious about is - any difference between the scores computed, is there some difference between the scores computed with this patch/without this patch?(IIRC, last time the issue was related to too many nodes having same score).

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/priorities/interpod_affinity.go]: @ravisantoshgudimetla could you point me the issue #?

This PR introduced following changes related with fraction:

- `weight` is always an integer, so new change that loads `weight` into `p.counts[node.Name]` (from float64 to int64) in an atomic way should be fine.
- change from `float64(term.Weight*int32(multiplier)))` to `int64(term.Weight*int32(multiplier))` should be also fine as `multiplier` is of type `int`
- `maxCount/minCount` are assigned from `p.counts[node.Name]`; should be good as well
- the last one: change from `(pm.counts[node.Name] - minCount) / (maxCount - minCount)` to `float64(*pm.counts[node.Name]-minCount) / float64(maxCount-minCount)`; this doens't lose correctness either.

**Comment by Huang-Wei** [comment]
/hold
in case of merge without thorough discussion.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go
Comment (bsalamat) [inline pkg/scheduler/algorithm/priorities/interpod_affinity.go]: why does the value need to be an `int64` pointer, instead of just an `int64`?

**Review by bsalamat — COMMENTED** [review]
Overall looks good and makes sense. Just a small comment.

**Comment by bsalamat** [comment]
Performance improvement is over 2X for preferred affinity. This is worth noting in the release notes.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/priorities/interpod_affinity.go]: It's because [atomic#AddInt64](https://golang.org/pkg/sync/atomic/#AddInt64) takes pointer as parameter.

If we go with `map[string]int64`, map value is not addressable (`&map["key"]` is illegal); and alsoyou can't make `val := map["key"]; ptr := &val` b/c that's another address.

**Comment by Huang-Wei** [comment]
@bsalamat Will update the release note.

One thing to note is that the 2X improvement is not for soft pod affinity, it's for **hard** pod affinity.

BTW: performance for soft(preferred) pod (anti-) affinity hasn't been measured due to lack of benchmark tests. But from the code's perspective, it's expected to have a performance improvement as well.

Why the priority changes can impact hard pod affinity is because of `hardPodAffinitySymmetricWeight`:

https://github.com/kubernetes/kubernetes/blob/552d1ebd7086dfd4ba5fdc73277866eb656eea0f/pkg/scheduler/algorithm/priorities/interpod_affinity.go#L164-L177

**Comment by bsalamat** [comment]
Thanks, @Huang-Wei for clarifying. You are right. This improvement impacts both soft and hard affinity.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go
Comment (bsalamat) [reply pkg/scheduler/algorithm/priorities/interpod_affinity.go]: That's right. In Go you cannot get a pointer to map entries as they may change during execution.

**Review by bsalamat — COMMENTED** [review]
/lgtm
/approve

Please change the release note and remove "hard" from it. This PR improves performance of both hard and soft affinity.

**Comment by Huang-Wei** [comment]
Yeap. And hard pod anti-affinity isn't involved in Priorities, so this PR doesn't help for that case.

BTW: I'm also trying to improve hard pod anti-affinity, such as:

1. change podsWithAffinity from slice to set 

https://github.com/kubernetes/kubernetes/blob/c4f97fa2a3a6f952cd2e0d93e237482990b13fe7/pkg/scheduler/nodeinfo/node_info.go#L52

2. split podsWithAffinity into podsWithAffinity and podsWithHardAntiAffinity, so as to short circuit:

https://github.com/kubernetes/kubernetes/blob/c4f97fa2a3a6f952cd2e0d93e237482990b13fe7/pkg/scheduler/algorithm/predicates/metadata.go#L402

But I don't see either of them gains performance improvement yet. Will continue digging.

**Comment by bsalamat** [comment]
One thing that can improve performance, is to remove `podsWithAffinity []*v1.Pod` from the `NodeInfo` and make it independent (or maybe a part of metadata). Today, we go over the whole "NodeInfo" array and check if any of them has a non-zero length `podWithAffinity`. In clusters with only a few affinity pods,  `podWithAffinity` is empty for most of the "NodeInfo" entries. If `podWithAffinity` was outside of NodeInfo, we could populate the entries only for those nodes which had a pod with affinity.

something like:
```go
NodesWithAffinityPods map[string]podsWithAffinity     // map from node name to affinity pods
```

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go
Comment (ravisantoshgudimetla) [reply pkg/scheduler/algorithm/priorities/interpod_affinity.go]: Got it, thanks for the explanation. I have some questions on the `term.Weight*int32(multiplier)` but I see that both of them are of the type int.

**Comment by Huang-Wei** [comment]
FYI: after merge of this PR, the perf improvement can be observed in perf.k8s.io:

image

**Comment by bsalamat** [comment]
@Huang-Wei Scheduling benchmarks show faster execution, but our Kubemark and real cluster benchmarks show a regression:
* Kubemark: http://perf-dash.k8s.io/#/?jobname=kubemark-5000Nodes&metriccategoryname=Scheduler&metricname=SchedulingLatency&Operation=priority_evaluation
* Real cluster benchmark: http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=Scheduler&metricname=SchedulingThroughput

**Pull request merged**
